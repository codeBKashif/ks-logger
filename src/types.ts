export enum LOG_LEVEL {
  ERROR = "error",
  INFO = "info",
  DEBUG = "debug",
}

export interface FormatterFact {
  date: string;
  loglevel: LOG_LEVEL;
  transactionId: string;
  error: string;
}

export interface FormatterClosureOptions {
  key: string;
  value: string | number | null | undefined;
}

export interface ParserInputInfo {
  date: string;
  level: LOG_LEVEL | string;
  data: {
    transactionId: string;
    details: string;
    code?: number;
    err?: string;
  };
}
