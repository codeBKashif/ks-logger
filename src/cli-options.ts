import { parse } from "ts-command-line-args";

interface CLIOptions {
  input: string;
  output: string;
  level?: string;
  help?: boolean;
}

export const getCliOptions = () => {
  return parse<CLIOptions>(
    {
      input: {
        type: String,
        multiple: false,
        description: "input file path",
      },
      output: {
        type: String,
        multiple: false,
        description: "output file path",
      },
      level: {
        type: String,
        multiple: false,
        optional: true,
        defaultValue: "error",
        description: "log level can be error, info, and debug",
      },
      help: { type: Boolean, optional: true, alias: "h", description: "Logger app usgae" },
    },
    {
      helpArg: "help",
      headerContentSections: [{ header: "Logger app" }],
    }
  );
};
