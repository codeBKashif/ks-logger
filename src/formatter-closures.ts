import { FormatterClosureOptions, FormatterFact } from "./types";

export const dateFormatter = (parsedInfo: FormatterFact): FormatterClosureOptions => {
  return { key: "timestamp", value: new Date(parsedInfo.date).getTime() };
};

export const transactionIdFormatter = (parsedInfo: FormatterFact): FormatterClosureOptions => {
  return { key: "transactionId", value: parsedInfo.transactionId };
};

export const logLevelFormatter = (parsedInfo: FormatterFact): FormatterClosureOptions => {
  return { key: "loglevel", value: parsedInfo.loglevel };
};

export const errorFormatter = (parsedInfo: FormatterFact): FormatterClosureOptions => {
  return { key: "err", value: parsedInfo.error };
};
