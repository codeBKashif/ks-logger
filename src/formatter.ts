import { FormatterFact } from "./types";

export class Formatter {
  private formatSequence: Function[];

  public constructor(sequence: Function[]) {
    this.formatSequence = sequence;
  }
  public format(formatterFact: FormatterFact): Record<string, string | number | null | undefined> {
    let result: Record<string, string | number | null | undefined> = {};
    this.formatSequence.forEach((sequence) => {
      const { key, value } = sequence(formatterFact);
      result[key] = value;
    });
    return result;
  }
}
