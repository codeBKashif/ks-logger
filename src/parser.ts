import { ParserInputInfo } from "./types";

export class Parser {
  public constructor() {}
  public parse(line: string): ParserInputInfo {
    const info = line.split(" - ");
    return {
      date: info[0],
      level: info[1],
      data: JSON.parse(info[2]),
    };
  }
}
