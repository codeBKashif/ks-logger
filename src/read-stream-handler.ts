import { createReadStream } from "fs";
import { createInterface, ReadLine } from "readline";

export class ReadStreamHandler {
  private readStream: ReadLine;

  public constructor(inputFile: string) {
    this.readStream = createInterface({
      input: createReadStream(inputFile),
      crlfDelay: Infinity,
    });
  }

  public getFileReader(): ReadLine {
    return this.readStream;
  }
}
