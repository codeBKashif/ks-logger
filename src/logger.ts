import { Formatter } from "./formatter";
import { Parser } from "./parser";
import { ReadStreamHandler } from "./read-stream-handler";
import { LOG_LEVEL, ParserInputInfo } from "./types";
import { WriteStreamHandler } from "./write-stream-handler";

export class Logger {
  private readStreamHandler: ReadStreamHandler;
  private parser: Parser;
  private formatter: Formatter;
  private writeStreamHandler: WriteStreamHandler;
  private outPutLevel: LOG_LEVEL;

  public constructor(
    readStreamHandler: ReadStreamHandler,
    writeStreamHandler: WriteStreamHandler,
    parser: Parser,
    formatter: Formatter,
    logLevel: LOG_LEVEL
  ) {
    this.readStreamHandler = readStreamHandler;
    this.writeStreamHandler = writeStreamHandler;
    this.parser = parser;
    this.formatter = formatter;
    this.outPutLevel = logLevel;
  }

  public log(): void {
    const outputData: Record<string, string | number | null | undefined>[] = [];
    this.readStreamHandler
      .getFileReader()
      .on("line", (line) => {
        const parsedData: ParserInputInfo = this.parser.parse(line);

        if (parsedData.level === this.outPutLevel) {
          outputData.push(
            this.formatter.format({
              date: parsedData.date,
              loglevel: parsedData.level as LOG_LEVEL,
              transactionId: parsedData.data.transactionId,
              error: parsedData.data.err || "",
            })
          );
        }
      })
      .on("close", () => {
        this.writeStreamHandler.writeData(JSON.stringify(outputData));
      });
  }
}
