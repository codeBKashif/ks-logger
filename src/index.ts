import { getCliOptions } from "./cli-options";
import { Formatter } from "./formatter";
import {
  dateFormatter,
  errorFormatter,
  logLevelFormatter,
  transactionIdFormatter,
} from "./formatter-closures";
import { Logger } from "./logger";
import { Parser } from "./parser";
import { ReadStreamHandler } from "./read-stream-handler";
import { LOG_LEVEL } from "./types";
import { WriteStreamHandler } from "./write-stream-handler";

const { input, output, level } = getCliOptions();

const readStreamHandler = new ReadStreamHandler(input);
const writeStream = new WriteStreamHandler(output);
const parser = new Parser();
const formatter = new Formatter([
  dateFormatter,
  transactionIdFormatter,
  logLevelFormatter,
  errorFormatter,
]);
const logger = new Logger(readStreamHandler, writeStream, parser, formatter, level as LOG_LEVEL);
logger.log();
