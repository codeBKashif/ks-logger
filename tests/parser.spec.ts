import { Parser } from "../src/parser";

describe("Parser", () => {
  let parser: Parser;

  beforeAll(() => {
    parser = new Parser();
  });

  it("should parse the line and return data", () => {
    const result = parser.parse(
      `2044-08-09T02:12:51.253Z - info - {"transactionId":"9abc55b2-807b-4361-9dbe-aa88b1b2e978","details":"Service is started"}`
    );
    expect(result).toMatchSnapshot();
  });
});
