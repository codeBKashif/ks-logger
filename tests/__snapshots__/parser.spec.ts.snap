// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser should parse the line and return data 1`] = `
Object {
  "data": Object {
    "details": "Service is started",
    "transactionId": "9abc55b2-807b-4361-9dbe-aa88b1b2e978",
  },
  "date": "2044-08-09T02:12:51.253Z",
  "level": "info",
}
`;
